### PR TITLE
add a specialized check for gradients (RequiresGradCheck) to speed up runtime assumptions check 

### PR DESCRIPTION
**Pull request by Krovatkin**
This change replaces general-case TypeChecks on DifferentiableGraphs addded in https://github.com/pytorch/pytorch/pull/49433 with specialized `RequiresGradCheck`s that only checks if the profiled `requires_gradients` properties match tensors' `requires_gard` at runtime. 
This change improves perf by 3-4% on fastrnns.

### CODE REVIEW

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp [hunk 1]
```diff
@@ -64,6 +65,32 @@ RegisterOperators reg(
            };
          },
          aliasAnalysisSpecialCase()),
+     Operator(
+         prim::RequiresGradCheck /* (...)  -> (..., bool) */,
+         [](const Node* node) -> Operation {
+           std::vector<c10::optional<bool>> rg_props =
```
Comment (bertmaher) [inline torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:71]: Just to be sure, this `std::vector` is only constructed once, right, and then cached inside the lambda below?  I just want to make sure we're not allocating on every RequiresGradCheck.

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp [hunk 2]
```diff
@@ -64,6 +65,32 @@ RegisterOperators reg(
...
+               fmap(node->tys(attr::types), [](const TypePtr& t) {
+                 return t->cast<TensorType>()->requiresGrad();
+               });
+           return [rg_props](Stack* stack) {
+             auto num_inputs = rg_props.size();
+             TORCH_INTERNAL_ASSERT(
+                 stack->size() >= num_inputs && num_inputs > 0);
```
Comment (bertmaher) [inline torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:78]: Any reason to check `num_inputs > 0` here?  Since this is a static property of the `RequiresGradCheck` op, let's catch it at construction time.  (I also have mixed feelings about `TORCH_INTERNAL_ASSERT` in the interpreter loop period, but I can set that aside for now ;-) )

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp [hunk 3]
```diff
@@ -64,6 +65,32 @@ RegisterOperators reg(
...
+             // Check every input's shape against profiled (expected) shape.
+             for (size_t i = 0; i < num_inputs; i++) {
+               auto& input = peek(stack, i, num_inputs);
+               auto t = input.toTensor();
```
Comment (bertmaher) [inline torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:82]: `auto t` will make a copy of the `Tensor`.  It seems like you should be able to use `auto const& t` but at least do `auto& t`.

**Review by bertmaher — COMMENTED** [review]
Test failures are probably related?

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:71]: yup!

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:78]: err, that was a copy-pasta from TypeCheck! I think I should update both.

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp (hunk 3 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:82]: ah! a good catch! fixed

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:71]: I  don't think rg-props  should ever contain c10::nullopt, that wouldn't be checking anything at all. (it also makes it so we have to do an extra unwrap of the bool). Can we change `std::vector<c10::optional<bool>> ` ->  `std::vector<bool>` ?

Comment (Krovatkin) [reply torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:70]: Yeah, we can, I just double check that Thomas indeed is conservatively setting `requiresGrad` to true if we it changed during profiling. Let me update it

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp [hunk 4]
```diff
@@ -64,6 +64,33 @@ RegisterOperators reg(
...
+               fmap(node->tys(attr::types), [](const TypePtr& t) {
+                 // if an rg property changes we assume a tensor does require
+                 // gradients which is set in `guardDifferentiableGraph`
+                 TORCH_INTERNAL_ASSERT(
+                     t->cast<TensorType>()->requiresGrad().has_value());
+                 return *t->cast<TensorType>()->requiresGrad();
+               });
+           return [rg_props](Stack* stack) {
+             auto num_inputs = rg_props.size();
+             // Check every input's shape against profiled (expected) shape.
```
Comment (eellison) [inline torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:80]: nit:  profile (expected) shape is copy-pasta

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp [hunk 5]
```diff
@@ -64,6 +64,33 @@ RegisterOperators reg(
...
+             for (size_t i = 0; i < num_inputs; i++) {
+               auto& input = peek(stack, i, num_inputs);
+               const auto& t = input.toTensor();
+               if (rg_props[i] != t.requires_grad()) {
```
Comment (eellison) [inline torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:84]: just wondering: is there any worry here about undefined tensor? (potentially not) what  happens when you call undefined.requires_grad() ?

File: torch/csrc/jit/passes/inline_autodiff_subgraphs.cpp [hunk 6]
```diff
@@ -21,7 +21,8 @@ bool canRunWithAutograd(Node* node) {
   }
   return kind != prim::FusionGroup && kind != prim::CudaFusionGroup &&
       kind != prim::TypeCheck && kind != prim::TensorExprGroup &&
-      kind != prim::CudaFusionGuard && (kind.is_aten() || kind.is_prim());
+      kind != prim::RequiresGradCheck && kind != prim::CudaFusionGuard &&
```
Comment (eellison) [inline torch/csrc/jit/passes/inline_autodiff_subgraphs.cpp:24]: why did you  add this here ?

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp (hunk 5 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:84]: undefined tensors shouldn't reach here: even with higher-order derivatives, we either guard `AutogradAdd` and `AutogradAnyNonZero` and similar ops or we don't differentiate them?

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp (hunk 4 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:80]: hehe, will fix.

File: torch/csrc/jit/passes/inline_autodiff_subgraphs.cpp (hunk 6 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/inline_autodiff_subgraphs.cpp]: you are right. it doesn't make too much sense since all RequiresGradCheck should be **outside** :-) or differentiable graphs

Comment (Krovatkin) [reply torch/csrc/jit/passes/inline_autodiff_subgraphs.cpp]: removed it.

**Review by eellison — APPROVED** [review]
nice !
